fairlearn/datasets/_fetch_acsincome.py

```diff
@@ -0,0 +1,143 @@
+        data : ndarray, shape (1664500, 10)
+            Each row corresponding to the 10 feature values in order.
+            If ``as_frame`` is True, ``data`` is a pandas object.
+        target : numpy array of shape (1664500,)
+            Integer denoting each person's annual income.
+            A threshold can be applied as a postprocessing step to frame this
+            as a binary classification problem.
+            If ``as_frame`` is True, ``target`` is a pandas object.
+        feature_names : list of length 10
+            Array of ordered feature names used in the dataset.
+        DESCR : string
+            Description of the ACSIncome dataset.
+
+    (data, target) : tuple of (numpy.ndarray, numpy.ndarray) or (pandas.DataFrame, pandas.Series)
+        if ``return_X_y`` is True and ``as_frame`` is False
+
+    (data, target) : tuple of (pandas.DataFrame, pandas.Series)
+        if ``return_X_y`` is True and ``as_frame`` is True
+
+    References
+    ----------
+    .. [1] Ding, F., Hardt, M., Miller, J., & Schmidt, L. (2021).
+       "Retiring Adult: New Datasets for Fair Machine Learning."
+       Advances in Neural Information Processing Systems, 34.
+
+    """
+    # State Code based on 2010 Census definitions
+    _STATE_CODES = {'AL': '01', 'AK': '02', 'AZ': '04', 'AR': '05', 'CA': '06',
+                    'CO': '08', 'CT': '09', 'DE': '10', 'FL': '12', 'GA': '13',
+                    'HI': '15', 'ID': '16', 'IL': '17', 'IN': '18', 'IA': '19',
+                    'KS': '20', 'KY': '21', 'LA': '22', 'ME': '23', 'MD': '24',
+                    'MA': '25', 'MI': '26', 'MN': '27', 'MS': '28', 'MO': '29',
+                    'MT': '30', 'NE': '31', 'NV': '32', 'NH': '33', 'NJ': '34',
+                    'NM': '35', 'NY': '36', 'NC': '37', 'ND': '38', 'OH': '39',
+                    'OK': '40', 'OR': '41', 'PA': '42', 'RI': '44', 'SC': '45',
+                    'SD': '46', 'TN': '47', 'TX': '48', 'UT': '49', 'VT': '50',
+                    'VA': '51', 'WA': '53', 'WV': '54', 'WI': '55', 'WY': '56',
+                    'PR': '72'}
+    if states is None:
+        states = _STATE_CODES.keys()
+    if not data_home:
+        data_home = pathlib.Path().home() / _DOWNLOAD_DIRECTORY_NAME
+
+    # fetch data for all 50 US states and Puerto Rico
+    data_dict = fetch_openml(
+        data_id=43141,
+        data_home=data_home,
+        cache=cache,
+        as_frame=True,
+        return_X_y=False,
+    )
+
+    # filter by state
+    df_all = data_dict['data'].copy(deep=True)
+    df_all['PINCP'] = data_dict['target']
+    cols = df_all.columns
+    df = pd.DataFrame(np.zeros((0, len(cols))), columns=cols)
+    for state in states:
+        dfs = [df, df_all.query(f"ST == {int(_STATE_CODES[state])}")]
+        df = pd.concat(dfs)
+    # drop the state column since it is not a feature in the published ACSIncome dataset
+    df.drop('ST', axis=1, inplace=True)
+
+    if as_frame:
+        data_dict['data'] = df.iloc[:, :10]
```

Comment:
What does the 10 represent? 

Relatedly, what is the team's policy on using magic numbers? @imatiach-msft

Reply:
10 refers to the number of features. It is hardcoded since the number of features is constant for this dataset. To improve readability, I'll update this by defining `_NUM_FEATS = 10` inside the function